### ISSUE TITLE
Timestamp unicity

### ISSUE THREAD

**Issue by leplatrem** [mozilla-services/syncto#82]
As [discussed with @michielbdejong](https://github.com/Kinto/kinto/pull/451/files#r54396093), there is a problem with the current synchronization implementation of Kinto.js and Sync data obtained via SyncTo:

> [... the timestamp unicity] constraint is not true, and records will actually get skipped when using this approach (in my test account, I had hundreds of records with identical timestamp, because the FxSync sets the first sync time as the timestamp for pre-existing bookmarks).

Should we add a decimal value on records in SyncTo so that timestamps are guaranteed to be unique? What can we do?

**Comment by Natim** [comment]
We could use the id that we know is unique to add a decimal value to existing timestamps.

**Comment by Natim** [comment]
This can be updated with 3 uniq digits https://github.com/mozilla-services/syncto/blob/master/syncto/views/record.py#L61

**Comment by Natim** [comment]
Something like `b64decode(id) & 0x3e7` should do it right?

**Comment by Natim** [comment]
Or a working code:

```
int(base64.b64decode(record['id']).encode('hex'), 16) & 0x3e7
```

**Comment by Natim** [comment]
@michielbdejong However it will work only if there are not a too big number of records per second.

**Comment by michielbdejong** [comment]
Oh, but this is just a documentation problem, right? It just means that with Syncto you shouldn't rely on incrementing the timestamp, and you should use proper `Next-Page` tokens, IIRC. See also https://github.com/Kinto/kinto.js/pull/299#issuecomment-169700244

**Comment by Natim** [comment]
We have validated that using the `Next-Page` the pagination works even if the records shares the same timestamp.

Also @rfk told me that at max 100 records can share the same timestamp (because the batch max size to add records is 100)

Which means that we can use [a collision model](http://preshing.com/20110504/hash-collision-probabilities/) to define the number of bytes that we need to take into account in order to be really unlikely to have collisions.

**Comment by Natim** [comment]
If we use less that 1/1000 collision probability and that we can use a 4 seconds window between two probable batch, then we could generate: timestamps like that: `1456825672037.8184` and are really likely to be unique for 100 records sharing the same timestamp. The mask would then be `0x4c4b3f`

Also the Kinto protocol doesn't really handles decimal values for timestamps so it would make it more complicate to handle.

Plus there are some timestamps that we would read without knowing what is the related id values. (Like the collection timestamp for instance)

We could say that instead of adding a value to the timestamps we would remove it so that the collection timestamp is always bigger than the record timestamps.

This seems to add a lot of point where it could fails. So maybe instead of betting on last_modified uniqueness for syncto we should make sure that we are using a synchronization strategy based on the `Next-Page` that comes directly from the Sync server pagination strategy.

**Comment by michielbdejong** [comment]
yeah, I would be in favour of adding a footnote to the kinto synchronization docs saying 'careful: this will not work in Syncto, you should use `Next-Page` tokens there'

**Comment by leplatrem** [comment]
:+1:

**Comment by rfk** [comment]
> at max 100 records can share the same timestamp (because the batch max size to add records is 100)

FWIW we're considering raising this limit in bugs like https://bugzilla.mozilla.org/show_bug.cgi?id=1250189 and https://bugzilla.mozilla.org/show_bug.cgi?id=1250747 although we haven't committed to anything specific.
